### ISSUE TITLE
Expanding AsyncLoopFactory to have startOnCondition

### ISSUE THREAD

**Issue by noescape00** [stratisproject/StratisBitcoinFullNode#948]
There are a lot of 'hacked' loops that would start executing only after n seconds. This is done to start the loop only when the feature\class this loop depends on initialized or finished doing something. 

Maybe it will be better to actually have a condition on when the loop can start, a condition that would be checked every ~0.1 seconds instead of assuming that target condition will be satisfied in constant X seconds? 

I suggest adding an overload for the IAsyncLoopFactory.Run that has startOnCondition of type Task<bool> as an input and refactoring loops that have 'startAfter' value set in order to wait for something else to load or execute. 

Also there are loops that have startAfter value the same as repeatEvery in order to start the loop with the delay. This can be done without specifying 2 timespans as different values (they are supposed to be the same in some cases, why having 2 different variables?)- instead we could add another overload like 
` IAsyncLoop Run(string name, Func<CancellationToken, Task> loop, CancellationToken cancellation, TimeSpan? repeatEvery = null, bool startWithDelay = false);` that would wait before running the first loop iteration for 'repeatEvery' seconds if 'startWithDelay' is 'true'.

**Comment by quantumagi** [comment]
Instead of having a `repeatEvery` maybe it would be better to specify a wait period between runs - i.e. `delayDuration`? I am thinking of avoiding scenarios where the repeatEvery may be less than the time the task takes to run once.

**Comment by quantumagi** [comment]
Perhaps:

`IAsyncLoop Run(string name, Func<CancellationToken, Task> loop, CancellationToken cancellation, TimeSpan? repeatEvery = null, Action<CancellationToken> startCondition = null);`

where startCondition can accept the result of one of these:

```
        private Action<CancellationToken> GetTaskAwaiter(Task task)
        {
            return (cancellation) =>
            {
                while (!cancellation.IsCancellationRequested)
                {
                    if (task.IsCompleted)
                        break;

                    Task.Delay(100, cancellation).ConfigureAwait(false);
                }
            };
        }

        private Action<CancellationToken> GetDelayAwaiter(TimeSpan delay)
        {
            return (cancellation) => Task.Delay(delay, cancellation);
        }
```

**Comment by Aprogiena** [comment]
conceptually, never do active wait, that's just no-go in almost every situation

also, i can't see any problems with specification of the two variables - when to start and the frequency

in general it might be a good idea to add start on condition (not with active wait though), but the question is - is it worth? so to justify such extension, i'd like to see couple of examples where you think this is actually an issue

to me, this sounds like a super low priority thing that maybe might not be worth even doing, but try to convince me with examples

**Comment by noescape00** [comment]
@Aprogiena actually yeah, this seems a very low priority fix to me now. Also I've checked usages of AsynLoop that could perform from this and it looks like it's better to just refactore those loops instead of overloading them with active waiting. 
Therefore I'm closing this issue.
